internal/controllers/machinedeployment/mdutil/util.go

```diff
@@ -469,58 +475,71 @@ func FindNewMachineSet(deployment *clusterv1.MachineDeployment, msList []*cluste
 		if upToDate {
-			matchingMachineSets = append(matchingMachineSets, ms)
+			newMSCandidates = append(newMSCandidates, ms)
 		} else {
-			diffs = append(diffs, fmt.Sprintf("MachineSet %s: diff: %s", ms.Name, strings.Join(logMessages, ", ")))
+			oldMSs = append(oldMSs, ms)
+			oldMSNotUpToDateResults[ms.Name] = *notUpToDateResult
+			diffs = append(diffs, fmt.Sprintf("MachineSet %s: diff: %s", ms.Name, strings.Join(notUpToDateResult.LogMessages, ", ")))
 		}
 	}
 
-	if len(matchingMachineSets) == 0 {
-		return nil, fmt.Sprintf("couldn't find MachineSet matching MachineDeployment spec template: %s", strings.Join(diffs, "; ")), nil
+	if len(newMSCandidates) == 0 {
+		return nil, oldMSs, oldMSNotUpToDateResults, fmt.Sprintf("couldn't find MachineSet matching MachineDeployment spec template: %s", strings.Join(diffs, "; "))
 	}
 
 	// If RolloutAfter is not set, pick the first matching MachineSet.
 	if deployment.Spec.Rollout.After.IsZero() {
-		return matchingMachineSets[0], "", nil
+		for _, ms := range newMSCandidates[1:] {
+			oldMSs = append(oldMSs, ms)
+			oldMSNotUpToDateResults[ms.Name] = NotUpToDateResult{
+				// No need to set log or condition message for discarded candidates, it is not used anywhere.
+				// NOTE: in future we might consider to move not selected newMSCandidates to newMS without performing rollouts, but this is not in scope for the first iteration of in-place.
+				EligibleForInPlaceUpdate: false,
+			}
+		}
+		return newMSCandidates[0], oldMSs, oldMSNotUpToDateResults, ""
 	}
 
 	// If reconciliation time is before RolloutAfter, pick the first matching MachineSet.
 	if reconciliationTime.Before(&deployment.Spec.Rollout.After) {
-		return matchingMachineSets[0], "", nil
+		for _, ms := range newMSCandidates[1:] {
+			oldMSs = append(oldMSs, ms)
+			oldMSNotUpToDateResults[ms.Name] = NotUpToDateResult{
+				// No need to set log or condition for discarded candidates, it is not used anywhere.
+				// NOTE: in future we might consider to move not selected newMSCandidates to newMS without performing rollouts, but this is not in scope for the first iteration of in-place.
+				EligibleForInPlaceUpdate: false,
+			}
+		}
+		return newMSCandidates[0], oldMSs, oldMSNotUpToDateResults, ""
 	}
 
 	// Pick the first matching MachineSet that has been created at RolloutAfter or later.
-	for _, ms := range matchingMachineSets {
-		if ms.CreationTimestamp.Sub(deployment.Spec.Rollout.After.Time) >= 0 {
-			return ms, "", nil
+	for _, ms := range newMSCandidates {
+		if newMS == nil && ms.CreationTimestamp.Sub(deployment.Spec.Rollout.After.Time) >= 0 {
+			newMS = ms
+			continue
+		}
+
+		oldMSs = append(oldMSs, ms)
+		oldMSNotUpToDateResults[ms.Name] = NotUpToDateResult{
+			// No need to set log or condition for discarded candidates, it is not used anywhere.
+			// NOTE: in future we might consider to move not selected newMSCandidates to newMS without performing rollouts, but this is not in scope for the first iteration of in-place.
```

Comment:
I think this comment should be different here. This is about a MS being created before rolloutAfter (which we said is entirely out of scope for in-place because it doesn't make sense)

Reply:
dropped (it is also explained in the comment added for https://github.com/kubernetes-sigs/cluster-api/pull/12811#discussion_r2394741498)